### ISSUE TITLE
Training with Soft Labels

### ISSUE THREAD

**Issue by SiftingSands** [facebookresearch/ClassyVision#618]
I was looking into using soft labels (e.g. [0.2, 0.3, 0.5] instead of one-hot encoding on the last class). However, I wanted to ask if there would be any major blockers for training in this manner. 

I know that I'll need to modify the loss and dataloader to handle a list of scores instead of a single class label. I'm not sure if that'll break other functions that expect the "[input, target]" default output from the dataloader (in the "_get_typed_sample" function).

I'm much less clear if I'll need to modify the training hooks or if other wrapper functions will cause issues.

Appreciate any advice and thanks!

**Comment by mannatsingh** [comment]
@SiftingSands this is already supported in Classy - you should be able to use [`SoftTargetCrossEntropyLoss`](https://classyvision.ai/api/losses.html#classy_vision.losses.SoftTargetCrossEntropyLoss) or [`LabelSmoothingCrossEntropyLoss`](https://classyvision.ai/api/losses.html#classy_vision.losses.LabelSmoothingCrossEntropyLoss) if you also want to use label smoothing. It should work out of the box, but if you run into any issues, let me know!

**Comment by SiftingSands** [comment]
Thanks for the fast response! I didn't even realize Classy already had soft targets implemented (My first thought was to just use PyTorch's KL divergence loss). I'll keep this ticket open in-case I run into any issues over the next week trying this out.

**Comment by SiftingSands** [comment]
I encountered one snag in the dataloader and SoftTargetCrossEntropyLoss interaction.

Tracing into the "forward" function in **soft_target_cross_entropy_loss.py** shows my **output** with the expected size of [batch_size, N_classes]. However, the **target** is a list of length "N_classes" and every element is a tensor of size "batch_size" (it's also a 64bit float, can't remember if that should be 32bit or not). 

This then fails the assert when it checks for shape matching between **target** and **output**. I'm assuming that I need to format my data differently, but I wanted to check with you for any suggestions before I dig through the abstractions.

Relevant parts of my dataloader (this scheme worked before for hard labels w/ torch.nn.CrossEntropyLoss()) :
``` 
def _get_typed_sample(input, target, sample_type):
    if sample_type == SampleType.DICT: # using this per default
        return {"input": input, "target": target}
    elif sample_type == SampleType.TUPLE:
        return (input, target)
    elif sample_type == SampleType.LIST:
        return [input, target]
    else:
        raise TypeError("Provided sample_type is not dict, list, tuple")

/////////other stuff in between////////////

def __getitem__(self, idx):
    # every sample is [filepath, label]
    image = Image.open(self.filepaths[idx][0]).convert('RGB')
    label = self.filepaths[idx][1] #list of class scores
        
    return _get_typed_sample(image, label, self.sample_type)
```

**Comment by mannatsingh** [comment]
The docs in https://classyvision.ai/api/losses.html#classy_vision.losses.SoftTargetCrossEntropyLoss.forward do mention the expected input formats - your target should be of the same shape as the model output.

If in doubt, you can always just initialize the loss in a notebook and try passing tensors of expected dimensions to make sure it works :)

**Comment by SiftingSands** [comment]
The fix was simply converting my label from a list to a tensor in ```__getitem__```. Closing as everything seems to be working with soft labels.
